torch_glow/src/FusingOptimizer.cpp

```diff
@@ -68,4 +73,106 @@ graph(%input, %weight):
   matmul_to_linear.runOnGraph(graph);
 }
 
+const torch::jit::Symbol& getGlowSymbol() {
+  static torch::jit::Symbol glow_sym = torch::jit::Symbol:fromQualString("glow::CompilationGroup");
+  return glow_symbol;
+}
+
+value_list sortReverseTopological(ArrayRef<Value*> inputs, Block* block) {
+  value_list result;
+  for (auto i : inputs) {
+    if (i->node()->owningBlock() == block) {
+	  result.push_back(i);
+	}
+  }
+
+  std::sort(result.begin(), result.end(), [&](Value* a, Value* b) {
+    return a->node()->isAfter(b->node());
+	});
+  return result;
+}
+
+bool canMerge(jit::torch::Node* node, isSupportFunc fn) {
+  return node->kind == torch::jit::prim::Constant || fn(node);
+}
+
+bool canMerge(jit::torch::Block* block, isSupportFunc fn) {
+  for (torch::jit::Node* node : block->nodes()) {
+    if (!canMerge(node, fn) {
+	  rerturn false;
+	}
+  }
+  return true;
+}
+
+#define REQ(cond)							\
+  if (!(cond)) {							\
+    llvm:err("Failed cond " #cond "\n");	\
+    return c10:nullptr;						\
+  }
+
+torch::c10::optional<torch::jit::Node*> tryMerge(torch::jit::Node* consumer, torch::jit::Node* producer, AliasDb& aliasdb, isSupportFunc fn) {
+
+  REQ(canMerge(producer, fn));
+  REQ(consumer->kind() == getGlowSymbol() || canMerge(consumer, aliasDb));
+
+  // Alias checks
+  // Requirement:
+  // - moveAfterTopologicallyValid(consumer, producer)
+  // - One of:
+  //   1) Both are in-place ops
+  //   2) Consumer is in-place, producer !hasInputWriters
+  //   3) Producer is in-place, consumer !hasOutputWriters
+  REQ(aliasDb.moveAfterTopologicallyValid(consumer, producer));
```

Comment:
it looks like this actually moves the consumer after producer, is there a reason we want to do this and not just see if it's possible with `couldMoveAfterTopologically`

Reply:
@bwasti

Reply:
Guess there is not